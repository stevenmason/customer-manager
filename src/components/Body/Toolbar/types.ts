import { setFilter, addNewEditingCustomer } from "../../../store/actions";

export interface Props {
  filter: string;
  setFilter: typeof setFilter;
  addNewEditingCustomer: typeof addNewEditingCustomer;
}
