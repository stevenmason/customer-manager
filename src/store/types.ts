import {
  setCustomers,
  setLoading,
  setCustomer,
  setEditingCustomer,
  setEditingCustomerValue,
  setFilter,
  deleteCustomer
} from "./actions";
import rootReducer from "./reducer";

export interface Customer {
  id: string;
  firstName: string;
  lastName: string;
  dateOfBirth: string;
}

export interface KeyedCustomers {
  [key: string]: Customer;
}

export type State = ReturnType<typeof rootReducer>;

export type Actions =
  | ReturnType<typeof setCustomers>
  | ReturnType<typeof setCustomer>
  | ReturnType<typeof deleteCustomer>
  | ReturnType<typeof setEditingCustomer>
  | ReturnType<typeof setEditingCustomerValue>
  | ReturnType<typeof setLoading>
  | ReturnType<typeof setFilter>;
