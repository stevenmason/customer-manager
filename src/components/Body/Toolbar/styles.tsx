import styled from "styled-components";

const Container = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  flex-wrap: wrap;
  box-sizing: border-box;
  width: 100%;
  background-color: #f1f1f1;
  padding: 10px;
`;

export { Container };
