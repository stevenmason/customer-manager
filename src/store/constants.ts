enum ACTIONS {
  SET_CUSTOMERS = "SET_CUSTOMERS",
  SET_CUSTOMER = "SET_CUSTOMER",
  DELETE_CUSTOMER = "DELETE_CUSTOMER",
  SET_EDITING_CUSTOMER = "SET_EDITING_CUSTOMER",
  SET_EDITING_CUSTOMER_VALUE = "SET_EDITING_CUSTOMER_VALUE",
  SET_LOADING = "SET_LOADING",
  SET_FILTER = "SET_FILTER"
}

const DEFAULT_CUSTOMER = {
  id: "-1",
  firstName: "",
  lastName: "",
  dateOfBirth: ""
};

export { ACTIONS, DEFAULT_CUSTOMER };
