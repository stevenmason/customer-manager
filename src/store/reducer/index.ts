import { combineReducers } from "redux";

import customers from "./customers";
import editingCustomer from "./editingCustomer";
import filter from "./filter";
import loading from "./loading";

const rootReducer = combineReducers({
  customers,
  filter,
  editingCustomer,
  loading
});

export default rootReducer;
