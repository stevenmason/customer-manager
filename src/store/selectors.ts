import { createSelector } from "reselect";

import { State, Customer } from "./types";
import { customerContainsFilter } from "./utils";

const customers = (state: State) => state.customers;
const loading = (state: State) => state.loading;
const filter = (state: State) => state.filter;
const editingCustomer = (state: State) => state.editingCustomer;

const filteredCustomers = createSelector(
  [customers, filter],
  (customers, filter) =>
    Object.values(customers).filter((customer: Customer) =>
      customerContainsFilter(customer, filter)
    )
);

export { filteredCustomers, loading, filter, editingCustomer };
