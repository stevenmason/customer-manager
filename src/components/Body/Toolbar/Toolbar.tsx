import React from "react";

import * as SC from "./styles";
import { Props } from "./types";
import Button from "../../../patterns/Button";
import Input from "../../../patterns/Input";

const Toolbar: React.FC<Props> = ({
  filter,
  setFilter,
  addNewEditingCustomer
}) => {
  return (
    <SC.Container>
      <Input label="Filter" name="filter" value={filter} onChange={setFilter} />
      <Button onClick={addNewEditingCustomer}>Add Customer</Button>
    </SC.Container>
  );
};

export default Toolbar;
