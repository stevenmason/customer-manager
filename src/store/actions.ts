import { uniqueId } from "lodash";

import { ACTIONS, DEFAULT_CUSTOMER } from "./constants";
import { Customer } from "./types";

const setCustomers = (customers: Customer[]) => ({
  type: ACTIONS.SET_CUSTOMERS as ACTIONS.SET_CUSTOMERS,
  customers
});

const setCustomer = (customer: Customer) => ({
  type: ACTIONS.SET_CUSTOMER as ACTIONS.SET_CUSTOMER,
  customer
});

const deleteCustomer = (id: string) => ({
  type: ACTIONS.DELETE_CUSTOMER as ACTIONS.DELETE_CUSTOMER,
  id
});

const setEditingCustomer = (editingCustomer: Customer | null) => ({
  type: ACTIONS.SET_EDITING_CUSTOMER as ACTIONS.SET_EDITING_CUSTOMER,
  editingCustomer
});

const addNewEditingCustomer = () => ({
  type: ACTIONS.SET_EDITING_CUSTOMER as ACTIONS.SET_EDITING_CUSTOMER,
  editingCustomer: {
    ...DEFAULT_CUSTOMER,
    id: uniqueId()
  }
});

const setEditingCustomerValue = (value: string, name: string) => ({
  type: ACTIONS.SET_EDITING_CUSTOMER_VALUE as ACTIONS.SET_EDITING_CUSTOMER_VALUE,
  name,
  value
});

const setLoading = (loading: boolean) => ({
  type: ACTIONS.SET_LOADING as ACTIONS.SET_LOADING,
  loading
});

const setFilter = (filter: string) => ({
  type: ACTIONS.SET_FILTER as ACTIONS.SET_FILTER,
  filter
});

export {
  setCustomers,
  setCustomer,
  setEditingCustomer,
  setEditingCustomerValue,
  setLoading,
  setFilter,
  addNewEditingCustomer,
  deleteCustomer
};
