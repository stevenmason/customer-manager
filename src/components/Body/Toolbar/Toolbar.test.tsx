import React from "react";
import renderer from "react-test-renderer";

import Toolbar from "./Toolbar";
import { setFilter, addNewEditingCustomer } from "../../../store/actions";

const PROPS = {
  filter: "",
  addNewEditingCustomer,
  setFilter
};

it("renders correctly", () => {
  const tree = renderer.create(<Toolbar {...PROPS} />);
  expect(tree.toJSON()).toMatchSnapshot();
});
