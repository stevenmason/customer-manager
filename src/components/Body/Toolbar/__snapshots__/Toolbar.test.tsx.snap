// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bdVaJa fzbFad"
>
  <div
    className="sc-htpNat zrlvf"
    size="LARGE"
  >
    <label
      className="sc-bxivhb ctASUB"
    >
      Filter
    </label>
    <input
      className="sc-ifAKCX jOZwNh"
      name="filter"
      onChange={[Function]}
      value=""
    />
  </div>
  <button
    className="sc-bwzfXH ppMEU"
    onClick={[Function]}
  >
    Add Customer
  </button>
</div>
`;
